client/appengine_devices.go

```diff
@@ -35,14 +35,45 @@ const (
 	AstarteDeviceAlias
 )
 
-// ListDevices returns a list of Devices in the Realm
+// ListDevices returns the list of Device IDs for all Devices in the Realm. The
+// returned result can be large, GetDeviceListPaginator can be used instead to
+// retrieve the device list incrementally.
 func (s *AppEngineService) ListDevices(realm string) ([]string, error) {
-	callURL, _ := url.Parse(s.appEngineURL.String())
-	callURL.Path = path.Join(callURL.Path, fmt.Sprintf("/v1/%s/devices", realm))
-	deviceList := []string{}
-	err := s.client.genericJSONDataAPIGET(&deviceList, callURL.String(), 200)
+	result := []string{}
+
+	paginator, err := s.GetDeviceListPaginator(realm, defaultPageSize)
+	if err != nil {
+		return result, err
+	}
+
+	for hasNext := paginator.HasNextPage(); hasNext; hasNext = paginator.HasNextPage() {
```

Comment:
Not sure the postcondition is actually needed here. Since the `GetNextPage` is embedded in the for-loop body I think the `hasNext = paginator.HasNextPage()` can be dropped, i.e.:
`for hasNext := paginator.HasNextPage(); hasNext; {`